### ISSUE TITLE
Allow null value in Maps

### ISSUE THREAD

**Issue by nebhale** [immutables/immutables#349]
Currently builder methods for `Map` types have implementations that prevent both `null` keys _and_ `null` values.  These implementations are a variant on the following:

```
public final Builder metadata(String key, Object value) {
  this.metadatas.put(
      Objects.requireNonNull(key, "metadatas key"),
      Objects.requireNonNull(value, "metadatas value"));
  return this;
}
```

I believe that `null` keys are fine to reject, but that `null` values should be allowed.  `null` values are often used to indicate the "removal" of an entry when used within JSON payloads.  Today, it's nearly impossible to handle the fact that some external actor is returning a JSON payload with `null` values.

**Comment by elucash** [comment]
Thank you for the request. It's hard to get the semantic and safety right in this case.
Do you have any proposals on how we can do this without just allowing `null` values and dropping some null-safety guarantees (as users find the null-safety guarantees as one of the major value proposition in Immutables toolkit) ? It will definitely will not work for Guava immutable collections, only when backed by JDK collections. Possible solution is to have style/feature flag to enable null values. Annotating with `Map<Key, @Nullable Value>` will not work as type annotations are not properly/portably handled for the nested generic types in the Annotation Processing API (to be clear, it's not a problem of the APIs, but rather bugs in Javac and ECJ compilers).

As a possible workaround, you can consider of using `Map<Key, Optional<Value>>` which is more verbose, but, arguably, might be a "honest" depiction of the desired structure. Such values are expected to be `null` when serialized to JSON.

**Comment by Zteve** [comment]
Further to Ben's request, and in support of it, I would like the _option_ to allow `null` elements of a collection (set with a singular attribute name) to be ignored rather than actually inserted in the collection.

For example:
```
    /**
     * Adds elements to {@link _ListRoutesRequest#getPaths() paths} list.
     * @param elements An array of paths elements
     * @return {@code this} builder for use in a chained invocation
     */
    public final Builder path(String... elements) {
      if (this.paths == null) {
        this.paths = new ArrayList<String>();
      }
      for (String element : elements) {
        this.paths.add(Objects.requireNonNull(element, "paths element"));
      }
      return this;
    }
```
would be replaced (when an option is provided `@IgnoreNullElements` ??) by something like this:
```
    public final Builder path(String... elements) {
      for (String element : elements) {
        Optional.ofNullable(element)
            .map(realElement -> this.paths().add(realElement));
      }
      return this;
    }

    private List<String> paths() {
      if (this.paths == null) {
        this.paths = new ArrayList<String>();
      }
      return this.paths;
    }
```

At the moment `@Nullable` elements retrieved from one interface cannot be passed one-by-one to another interface in a collection, without considerable effort (just think of two of them!).

**Comment by elucash** [comment]
I would be glad to come up with a solution. Currently, I see the possibility of introducing
`@NullElements` attribute annotation which will work to make collection elements and map values to allow null elements. `@NullElements(ignore = true)` might be used to ignore null values.
If you will use it all over the place, there's an option to have it as a style/feature flag: it would be somewhat easier to implement it this way.

Could you share more info about the use cases? It's seems that I not fully understand the need behind this and the usage seems to be quite exotic. There's possibility that some really flexible and more general solution might be applied to customize checking/filtering of elements.

**Comment by Zteve** [comment]
In our context, we use Immutables pervasively for request and response classes: effectively allowing a flexible parameter passing and canonical response format for an API.  We use `null` values to denote the omission of optional attributes.

When building another layer on top of the API we have another set of request and response immutables, and the lower-level API is used to implement the higher level API.  In this case there is often the need to pass attributes from one layer to the other, and if they are optional in one layer they may also be omitted in the other.  This is fine if the attributes match one-for one—we simply pass the value on, `null` or not.  However, there are occasions where single element parameters from one layer become elements of collections in another.  In this case, we have to explicitly code for the omitted (`null`) case.  If several optionally omitted elements contribute to a single collection this can be a lot of code, especially if they are all omitted: when the collection itself may be `null`. It typically involves the construction of temporary intermediate mutable collections—something we had hoped the Immutables' builders would handle for us.

We observe that, with an "ignore null elements" option, this awkward coding would be unnecessary.

I imagine we could have "ignore null values" to mean: permit `null` values, but do nothing; "respect null values" to insert `null` values in the map/set/list &c.; and the default would be "error on null" (or something), which is the current behaviour.

For us, although we would probably want to use a fixed setting nearly everywhere, it would still be necessary to have an override for particular collection attributes.

I hope that helps.

**Comment by nebhale** [comment]
@elucash The summary provided by @Zteve is correct and complete, but at a higher level, we're trying to abide by [Postel's Law](https://en.wikipedia.org/wiki/Robustness_principle) and be liberal in what we accept.  I agree that generally, we might want to reject `null` values, but in situations where we're deserializing JSON payloads, it's important for us to take as much data as we can, regardless of its correctness.  Once deserialization happens, we can clean up that data, removing `null`s and the like, but as it stands now, there's a blocker to us accepting the data in the first place, as we use the Immutables generated objects for JSON deserialization.

In our particular implementation, it'd be important to preserve a key with a `null` value rather than not adding the key to the `Map` in the first place.  This is because, while `get(<KEY-WITH-NULL-VALUE>)` and `get(<KEY-DOES-NOT-EXIST>)` return the name `null` value and should be indistinguishable from one another, iteration of the the collection of `Entry`s is not the same.  When sending subsets of `Map`s serialized to JSON, it's a common idiom to use `key: null` to indicate removal, while the absence of a key in the map means "no change".  So I'm not sure that `ignore` is necessarily, the correct verb for configuration, and perhaps something like `@NullElements(allow = true)` is more accurate.

**Comment by Zteve** [comment]
@nebhale: Yes; and _three_ settings are desirable:

1. one to **_allow_** `null` values in `Collections`, that is in `List`s, `Set`s, and in the _range_ of a `Map`, &c., 
2. one to **_ignore_** (the setting of) `null` values in these places, and 
3. one to _**prevent**_ setting `null` in these places.

I'm not sure if the `null` value in the range of a `Map` isn't a special case, as it seems to hold a different amount of information to a `null` value in a `List`, for example. Perhaps we should only ignore `Map` singular settings if the **_key_** is `null` (that is, to redefine option 2.).

**Comment by elucash** [comment]
Does "range of a map" mean values of a map? Currently, I think than having null keys doesn't make much sense, so assume this is about values (like in `map.values()`).

By discussing the API options I want to reach a balance between clarity and minimalism.

The API might be as follows:
1. No annotation, default behavior, null-hostile, i.e. no nulls allowed
2. `@NullElements` allows null elements
3. `@NullElements(ignore = true)` to skip null elements

As an alternative, If java 8 is an option, we can try to use type annotation (despite complications I've mentioned before): `Map<String, @Nullable Value>` to allow nulls.

Ignoring nulls can be hardly considered a trait of type but rather a way an attribute is handled, so ignoring nulls might be better be a separate method-level annotation: `@IgnoreNulls Map<String, Value> attribute()`. `Style.ignoreNulls = true` feature flag may be used instead of `@IgnoreNulls` if it's always desired to skip nulls unless nulls are permitted by type annotation (as in  `Map<T, @Nullable V>`). This could lead us to reusing `@Nullable` annotation on a value type and one style attribute to skip nulls for non-nullable positions.

**Comment by Zteve** [comment]
Eugene,

Yes, by **_range_** I meant the set of "values" of a map. Similarly, **_domain_** is the set of "keys" of a map.

I think your proposal has merit, thank you.

**Comment by nebhale** [comment]
I quite like the `Map<T, @Nullable V>` style because it is very expressive with regards to the expected behavior.  As we don't have a need for a global "ignore" I don't have a clear idea of how useful it would be, but it certainly seems like it has merit.

**Comment by alexio** [comment]
If you do need a "null" value, an optional value sounds better than an nullable one in my opinion. `Map<T, Optional<V>>`

**Comment by elucash** [comment]
@alexio certainly, we discussed that variant, which is definitely a way to go and I, personally, would choose it over anything nullable at any time.
Per my understanding, the main motive here is an integration and pouring data back and forth with the sources of varying quality/consistency. Also many our users on Android found it prohibitive costly to use `Optional`s so resort to nullable. We already came an extra mile to bridge those needs so it is a reasonable idea to add null elements handling, but I would be reluctant to add anything else related to nulls afterwards ).

**Comment by digitalbuddha** [comment]
Additionally I've had the use case on Android were I extended an ORM's generated interface with immutables.  For this to work you would need the other library to support Optionals which isn't always guaranteed

**Comment by nebhale** [comment]
I fundamentally believe in `Optional`s and we do use them internally, but I have trouble subjecting outside users to them at this point.  Their status as a type rather than a language construct (á la Swift) makes them a difficult sell for API targeted at users who are not familiar with their behavior.  This is my only hesitation with the `Optional` pattern at this time, but maybe I just need to get over that and push my users along.

**Comment by elucash** [comment]
I've released this in the experimental forms in 2.2.5, available on Central.
There are some caveats of course!

First of all, I've used "ephemeral annotations" approach. You use annotations, but they are not provided by Immutables, so BYO@ (apparently I keep inventing funny terms). They are matched textually, so if you create `@SkipNulls` and `@Nullable` (or `@NullableElement` as it allows arbitrary suffix) in any package with `@Target(ElemenType.TYPE_USE)` and put them on types inside angular brackets (java 8 source level required). These are only for compilation, so don't have to redistribute the annotations.

The second is that there are no extensive validation of these annotations and combination of thereof, as such, any misuse might not be apparent right away.

As I mentioned it works only when no Guava immutable collections are used as they are always null-hostile: make sure there's no Guava in a classpath or use `Style.jdkOnly = true` to force JDK collections.

```java
@Value.Style(jdkOnly = true)
@Value.Immutable
public interface Elements {
  List<@Nullable Void> al();
  List<@SkipNulls String> sk();
  Map<String, @Nullable Integer> bl();
  Map<String, @SkipNulls Integer> sm();
}
```

**Comment by Zteve** [comment]
Just trying to get my head around `List<@Nullable Void>` :-D

This reads well and is pretty cool.

**Comment by digitalbuddha** [comment]
hi @elucash , I am using Immutables in Android and unfortunately cannot use Java8 yet.  Is there any other way that I can mark a `    List<Foo> getFoo(); as being able to have null values?  Thank you kindly.

**Comment by elucash** [comment]
Hi Mike, please see #452 , where we updated this functionality to work on attribute level
TL;DR
`@AllowNulls` and `@SkipNulls` would be recognized in 2.3.4

**Comment by digitalbuddha** [comment]
Sorry if unrelated, upgrading to 2.3.4 gives a Error:(13, 24) error: cannot find symbol class CheckReturnValue on Android. is there an additional dependency that needs to be added?

**Comment by elucash** [comment]
Let me see, In theory, it should not be used if `CheckReturnValue` is not on the classpath. Maybe it's a bug

**Comment by digitalbuddha** [comment]
I was able to solve the dependency with 
```    provided group: 'com.google.code.findbugs', name: 'jsr305', version: '1.3.9'```

But I am still having trouble getting the null to serialize, here's a more specific example.

given the following class
```
@Value.Immutable
public abstract class PuzzleData {
    
    public abstract List<PotentialRebusType> getAnswers();

}
```
I would like getAnswers to work with null values.

I created annotation 
```
@Documented
@Retention(RetentionPolicy.CLASS)
@Target({ElementType.METHOD, ElementType.PARAMETER, ElementType.FIELD})
public @interface AllowNulls {
}
```

Then I changed my class to have 
```
 @AllowNulls
    public abstract List<PotentialRebusType> getAnswers();
```

I am getting following error
```
java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:210)
	at com.google.common.collect.ImmutableCollection$ArrayBasedBuilder.add(ImmutableCollection.java:339)
	at com.google.common.collect.ImmutableList$Builder.add(ImmutableList.java:652)
	at com.nytimes.crossword.retrofit.ImmutablePuzzleData$Builder.addAnswers(ImmutablePuzzleData.java:443)
	at com.nytimes.crossword.retrofit.GsonAdaptersRetrofit$PuzzleDataTypeAdapter.readInAnswers(GsonAdaptersRetrofit.java:2004)
	at com.nytimes.crossword.retrofit.GsonAdaptersRetrofit$PuzzleDataTypeAdapter.eachAttribute(GsonAdaptersRetrofit.java:1933)
```

**Comment by elucash** [comment]
That's because Guava immutable collections do not allow nulls anyway, for this to work you need to switch to `Style(jdkOnly=true)`

**Comment by elucash** [comment]
@digitalbuddha JFYI, just released 2.3.5 which should not require `CheckReturnValue`
